### PR TITLE
8306042: C2: failed: Missed optimization opportunity in PhaseCCP (adding LShift->Cast->Add notification)

### PR DESCRIPTION
**Pull request by eme64**
An other case of `uncast` not being type-propagated through.

We have a case like this:
`Phi -> ShiftL -> CastII -> AndI`

The Phi has an updated type, so we should re-run Value on the AndI.

In PhaseCCP::push_and, we do update a similar pattern:
`X -> ShiftL -> AndI`

I extended it to handle this pattern:
`parent -> LShift (use) -> ConstraintCast* -> And`

For this, I implemented:
https://github.com/openjdk/jdk/blob/26f4adaae901822bea984b926c06d1a78f9c6b48/src/hotspot/share/opto/castnode.hpp#L73-L78

I could refactor code from a previous similar fix, for pattern: `ConstraintCast+ -> Sub/Phi`

**Discussion**

https://github.com/openjdk/jdk/blob/4d350f8f4eaabb18482c7656cb56a734e60187cf/src/hotspot/share/opto/castnode.hpp#L78-L79
I would have liked to place a `ResourceMark` between these two lines, to ensure the `internals` data structure is de-allocated after the traversal. But if I add it there, then one cannot modify any outer data-structure, or else one risks re-allocation of the outer data-structure in the inner ResourceMark, and then this memory gets de-allocated once the ResourceMark is cleared, and the outer data-structure is broken. This would for example mean that I could not push to the IGVN worklist inside the callback.

Not having the ResourceMark means a memory leak, until the compile phase is over. But my code is not the only place, there are lots of places where we create a Resource allocated data-structure, but do not use ResourceMarks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306042](https://bugs.openjdk.org/browse/JDK-8306042): C2: failed: Missed optimization opportunity in PhaseCCP (adding LShift-&gt;Cast-&gt;Add notification)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [4d2a4f9e](https://git.openjdk.org/jdk/pull/13611/files/4d2a4f9eaeab7d7c91b8ceb948a0e99da9429773)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13611/head:pull/13611` \
`$ git checkout pull/13611`

Update a local copy of the PR: \
`$ git checkout pull/13611` \
`$ git pull https://git.openjdk.org/jdk.git pull/13611/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13611`

View PR using the GUI difftool: \
`$ git pr show -t 13611`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13611.diff">https://git.openjdk.org/jdk/pull/13611.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13611#issuecomment-1521215828)

### CODE REVIEW

File: src/hotspot/share/opto/castnode.hpp [hunk 1]
```diff
@@ -69,6 +69,27 @@ class ConstraintCastNode: public TypeNode {
...
   Node* optimize_integer_cast(PhaseGVN* phase, BasicType bt);
+
+  // Visit all non-cast uses of the node, bypassing ConstraintCasts.
+  // Pattern: this (-> ConstraintCast)* -> non_cast
+  // In other words: find all non_cast nodes such that
+  // non_cast->uncast() == this.
+  template <typename Callback>
+  static void visit_uncasted_uses(const Node* n, Callback callback) {
+    Unique_Node_List internals;
+    internals.push((Node*)n); // start traversal
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/castnode.hpp:80]: We should get rid of these casts with [JDK-8252694](https://bugs.openjdk.org/browse/JDK-8252694). I added a comment there.

File: test/hotspot/jtreg/compiler/ccp/TestShiftCastAndNotification.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * @library /test/lib
+ * @run main/othervm -Xcomp -XX:-TieredCompilation
+ *                   -XX:CompileOnly=compiler.ccp.TestShiftCastAndNotification::test
+ *                   compiler.ccp.TestShiftCastAndNotification
+ */
+
+package compiler.ccp;
+
+public class TestShiftCastAndNotification {
+  static int N;
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/ccp/TestShiftCastAndNotification.java:37]: Should use 4-whitespace indentation.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I like the new `visit_uncasted_uses`.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by eme64** [comment]
@chhagedorn @TobiHartmann Thanks for the reviews.
I could integrate now.

But I have one more discovery:
The reason why I cannot place the mentioned `ResoureMark` is that the `worklist` has its arena as the `Thread::current()->resource_area()` in some cases. One example is the `Unique_Node_List worklist` in `PhaseCCP::analyze`. So then in `PhaseCCP::push_and`, this `worklist` gets captured by the lambda, and passed to `ConstraintCastNode::visit_uncasted_uses`. If there is a `ResourceMark` inside it, and we re-allocate memory for the `worklist` because it grows beyond its current capacity, then we will de-allocate that memory when we exit the `ResourceMark`, and the `worklist` now points to de-allocated memory.

Without a `ResourceMark`, this can be avoided. But then we never release the memory for the `Unique_Node_List internals`. 

The second use of `ConstraintCastNode::visit_uncasted_uses` in `PhaseIterGVN::add_users_to_worklist` does not run into this issue, even if we have a `ResourceMark`. The reason is that there the `worklist` that is captured allocates from a separate arena, namely the `Compile::current()->comp_arena()`.

We could thus also consider having the `worklist` inside `PhaseCCP::analyze` allocate also from the `Compile::current()->comp_arena()`. One downside is that we never de-allocate memory in that arena until the end of `Compile`. But currently, there is also no `ResourceMark` upstream of `PhaseCCP::analyze`, all the way up to `Compile`. So this `worklist` never has its memory de-allocated anyway.

What is your opinion? Should I make such changes, or just integrate what I have now?

**Review by TobiHartmann — APPROVED** [review]
The new changes look good to me. Let's clean up memory management of worklists with [JDK-8302670](https://bugs.openjdk.org/browse/JDK-8302670).

**Comment by eme64** [comment]
I now added the `worklist` to a local arena instead. I may change this again in [JDK-8302670](https://bugs.openjdk.org/browse/JDK-8302670), if that is beneficial for memory management of the CCP data-structures.

**Review by TobiHartmann — APPROVED** [review]

**Review by chhagedorn — APPROVED** [review]
Update looks good!

File: src/hotspot/share/opto/phaseX.cpp [hunk 3]
```diff
@@ -1965,8 +1950,13 @@ void PhaseCCP::analyze() {
...
+  // CCP worklist is placed on a local arena, so that we can allow ResourceMarks on "Compile::current()->resource_arena()".
+  // We also do not want to put the worklist on "Compile::current()->comp_arena()", as that one only gets de-allocated after
+  // Compile is over. The local arena gets de-allocated at the end of its scope.
+  ResourceArea local_arena(mtCompiler);
+  Unique_Node_List worklist(&local_arena);
+
   // Push root onto worklist
-  Unique_Node_List worklist;
   worklist.push(C->root());
   DEBUG_ONLY(Unique_Node_List worklist_verify;)
```
Comment (vnkozlov) [inline src/hotspot/share/opto/phaseX.cpp:1961]: Should you put `worklist_verify` to `local_arena` too?

Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp:1961]: @vnkozlov I could do that, but it is not required. The CCP `worklist` gets passed downward much farther, including the  graph walks in `push_child_nodes_to_worklist`. So there it is nice to be able to have `ResourceMarks`. But `worklist_verify` is only modified directly in `PhaseCCP::analyze` and one layer deeper in `PhaseCCP::verify_analyze`. So I would never expect a `ResourceMark` to mess with re-allocation.

So at this point the only reason to add `worklist_verify` to `local_arena` is to ensure it is de-allocated afterward. We could also use a `ResourceMark` and leave it on `Thread::current()->resource_area()`.

Anyway, I'll just move it to `local_arena`, after all it is `DEBUG_ONLY`.

Comment (vnkozlov) [reply src/hotspot/share/opto/phaseX.cpp]: ok

**Review by vnkozlov — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]
Still good. Ship it!

**Comment by eme64** [comment]
@vnkozlov @chhagedorn @TobiHartmann Thanks for the reviews and discussion!
/integrate

**Pull request closed without merging**